File: src/stb_image_write_wrapper.f90
module stb_image_write_wrapper
    !! Fortran interface to STB image write C wrapper
    !! Provides iso_c_binding interface to stb_image_write functions
    use iso_c_binding
    use, intrinsic :: iso_fortran_env, only: int8
    implicit none
    
    private
    public :: stb_write_jpeg_wrapper, stb_write_jpeg_to_memory
    
    ! C wrapper interfaces
    interface
        ! Write JPEG to file
        function stb_write_jpeg_wrapper(filename, width, height, components, data, quality) &
                bind(C, name="stb_write_jpeg_wrapper")
            import :: c_int, c_char, c_ptr
            character(c_char), intent(in) :: filename(*)
            integer(c_int), value :: width, height, components, quality
            type(c_ptr), value :: data
            integer(c_int) :: stb_write_jpeg_wrapper
        end function stb_write_jpeg_wrapper
        
        ! Write JPEG to memory buffer
        function stb_write_jpeg_to_memory(output_data, output_size, width, height, components, data, quality) &
                bind(C, name="stb_write_jpeg_to_memory")
            import :: c_int, c_ptr
            type(c_ptr), intent(out) :: output_data
            integer(c_int), intent(out) :: output_size
            integer(c_int), value :: width, height, components, quality
            type(c_ptr), value :: data
            integer(c_int) :: stb_write_jpeg_to_memory
        end function stb_write_jpeg_to_memory
    end interface
    
end module stb_image_write_wrapper

File: src/stb_image_write_wrapper.c
/**
 * STB Image Write C wrapper for fortplot
 * Provides simplified C interface to stb_image_write.h for JPEG output
 */

#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "../thirdparty/stb_image_write.h"
#include <stdio.h>
#include <stdlib.h>

// Wrapper for JPEG writing
int stb_write_jpeg_wrapper(const char* filename, int width, int height, int components, const void* data, int quality) {
    return stbi_write_jpg(filename, width, height, components, data, quality);
}

// Wrapper for in-memory JPEG writing
typedef struct {
    unsigned char* data;
    int size;
    int capacity;
} jpeg_buffer_t;

static void stb_write_callback(void *context, void *data, int size) {
    jpeg_buffer_t* buffer = (jpeg_buffer_t*)context;
    
    // Expand buffer if needed
    if (buffer->size + size > buffer->capacity) {
        buffer->capacity = (buffer->size + size) * 2;
        buffer->data = (unsigned char*)realloc(buffer->data, buffer->capacity);
    }
    
    // Copy data
    memcpy(buffer->data + buffer->size, data, size);
    buffer->size += size;
}

int stb_write_jpeg_to_memory(unsigned char** output_data, int* output_size, 
                             int width, int height, int components, const void* data, int quality) {
    jpeg_buffer_t buffer = {0};
    buffer.capacity = width * height * components + 1024; // Initial size estimate
    buffer.data = (unsigned char*)malloc(buffer.capacity);
    
    if (!buffer.data) return 0;
    
    int result = stbi_write_jpg_to_func(stb_write_callback, &buffer, width, height, components, data, quality);
    
    if (result) {
        *output_data = buffer.data;
        *output_size = buffer.size;
    } else {
        free(buffer.data);
        *output_data = NULL;
        *output_size = 0;
    }
    
    return result;
}

void stb_free_jpeg_memory(unsigned char* data) {
    free(data);
}